src/govuk/i18n.unit.test.mjs

```diff
@@ -39,4 +39,91 @@ describe('I18n', () => {
       expect(() => i18n.t()).toThrow('i18n lookup key missing.')
     })
   })
+
+  describe('string interpolation', () => {
+    const config = {
+      nameString: 'My name is %{name}'
+    }
+
+    it('throws an error if the options data is not present', () => {
+      const i18n = new I18n(config)
+      expect(() => { i18n.t('nameString') }).toThrowError('i18n: cannot replace placeholders in string if no option data provided')
+    })
+
+    it('throws an error if the options object is empty', () => {
+      const i18n = new I18n(config)
+      expect(() => { i18n.t('nameString', {}) }).toThrowError('i18n: no data found to replace %{name} placeholder in string')
+    })
+
+    it('throws an error if the options object does not have a matching key', () => {
+      const i18n = new I18n(config)
+      expect(() => { i18n.t('nameString', { unrelatedThing: 'hello' }) }).toThrowError('i18n: no data found to replace %{name} placeholder in string')
+    })
+
+    it('only matches %{} as a placeholder', () => {
+      const i18n = new I18n({
+        price: '%{name}, this } item %{ costs $5.00'
+      })
+      expect(i18n.t('price', { name: 'John' })).toBe('John, this } item %{ costs $5.00')
+    })
+
+    it('can lookup a placeholder value with non-alphanumeric key', () => {
+      const i18n = new I18n({
+        age: 'My age is %{current-age}'
+      })
+      expect(i18n.t('age', { 'current-age': 55 })).toBe('My age is 55')
+    })
+
+    it('can lookup a placeholder value with reserved name as key', () => {
```

Comment:
I think it might be worth either modifying this test or adding another one where the options object does _not_ have a key matching the name of the function in the prototype, as otherwise we wouldn't go up the prototype chain anyway.

(I've realised that `assign` is a static method on the Object class, and so wouldn't work for these examples, but `valueOf` would)

Basically I think we want to check that the way we're using `Object.hasOwnProperty` is doing its job and we'll never try and replace a property with something from the prototype chain.

If we were using `if (placeholderKey in options)` instead, then we could end up doing something like this:

```
> config = new Object
{}
> config.name = 'Ollie'
'Ollie'
> 'valueOf' in config
true
> 'I have the value of %{valueOf}'.replace('%{valueOf}', config.valueOf)
Uncaught TypeError: Cannot convert undefined or null to object
    at valueOf (<anonymous>)
    at String.replace (<anonymous>)
```

Let me know if that still doesn't make sense – finding it hard to explain!

Reply:
Added a new test here: https://github.com/alphagov/govuk-frontend/pull/2822/files#diff-ebbbe0f92d8344a9adf41d13a754cfddb236d06c1119a3b68226746800972cfdR84